text/0000-trait-alias.md

```diff
@@ -0,0 +1,166 @@
+our backend:
+
+```rust
+pub trait FooBackend: Foo<Backend> {
+  // ...
+}
+
+fn use_foo<A>(_: A) where A: FooBackend {}
+```
+
+If you try to pass an object that implements `Foo<Backend>`, that won’t work, because it doesn’t
+implement `FooBackend`. However, we can make it work with the following universal `impl`:
+
+```rust
+impl<T> FooBackend for T where T: Foo<Backend> {}
+```
+
+With that, it’s now possible to pass an object that implements `Foo<Backend>` to a function
+expecting a `FooBackend`. However, what about impl blocks? What happens if we implement only
+`FooBackend`? Well, we cannot, because the trait explicitely states that we need to implement
+`Foo<Backend>`. We hit a problem here. The problem is that even though there’s a compatibility at
+the `trait bound` level between `Foo<Backend>` and `FooBackend`, there’s none at the `impl` level,
+so all we’re left with is implementing `Foo<Backend>` – that will also provide an implementation for
+`FooBackend` because of the universal implementation just above.
+
+## Second example: ergonomic collections and scrapping boilerplate
+
+Another example is associated types. Take the following [trait from tokio](https://docs.rs/tokio-service/0.1.0/tokio_service/trait.Service.html):
+
+```rust
+pub trait Service {
+  type Request;
+  type Response;
+  type Error;
+  type Future: Future<Item=Self::Response, Error=Self::Error>;
+  fn call(&self, req: Self::Request) -> Self::Future;
+}
+```
+
+It would be nice to be able to create a few aliases to remove boilerplate for very common
+combinations of associated types with `Service`.
+
+```rust
+Service<Request = http::Request, Response = http::Response, Error = http::Error>;
+```
+
+The trait above is a http service trait which only the associated type `Future` is left to be
+implemented. Such an alias would be very appealing because it would remove copying the whole
+`Service` trait into use sites – trait bounds, or even trait impls. Scrapping such an annoying
+boilerplate is a definitive plus to the language and might be one of the most interesting use case.
+
+# Detailed design
+[design]: #detailed-design
+
+The syntax chosen to make a *trait alias* is:
+
+```rust
+trait TraitAlias = Trait;
+```
+It creates a new trait alias `TraitAlias` that will resolve to `Trait`.
+
+```rust
+trait DebugAlias = Debug;
+```
+
+Trait aliasing to combinations of traits is also provided with the standard `+` construct:
+
+```rust
+trait DebugDefault = Debug + Default;
+```
+
+Optionally, if needed, one can provide a `where` clause to express *bounds*:
+
+```rust
+trait DebugDefault = Debug where Self: Default; // same as the example above
+```
+
+Furthermore, it’s possible to use only the `where` clause by using the special syntax `= _`:
+
+```rust
+trait DebugDefault = _ where Self: Debug + Default;
+```
+
+Trait aliases can be used in any place arbitrary bounds would be syntactically legal. However, you
```

Comment:
In terms of structure, the detailed design could probably be split into two sections - syntax and semantics. Everything above this paragraph would be syntax, whereas this paragraph should be expanded into the semantics secion.

There are basically two cases semantically that we're allowing (and a third we're not - defining impls). The first is bounds, which is sort of straightforward, but has some question still around which bounds get elaborated as Niko discusses [here](https://github.com/rust-lang/rfcs/pull/1733#issuecomment-269219626). That is, in `trait Foo<T: Bar>: Baz<T>;`, do we enforce that `T: Bar`? I'm not sure what the answer is, @nikomatsakis probably should comment.

Objects are a bit trickier. You don't mention it, but an alias needs to be object safe in order for it to be used as an object. I believe this is the requirements for an alias to be object safe:

1. It contains zero or one object safe traits and zero or more of these other bounds: Send, Sync, 'static (that is, `trait Show = Display + Debug` would not be object safe).
2. All the associated types of the trait need to be specified.